src/structure/module.rs

```diff
@@ -59,3 +64,115 @@ pub trait VectorSpace<S: Field>
     : VectorSpaceApprox<S>
     + Module<S>
 {}
+
+
+/// A normed approximate vector space.
+pub trait NormedSpaceApprox<S: FieldApprox>
+    : VectorSpaceApprox<S> {
+    /// The squared norm of this vector.
+    fn norm_squared(&self) -> S;
+
+    /// The norm of this vector.
+    fn norm(&self) -> S;
+
+    /// Returns a normalized version of this vector.
+    fn normalize(&self) -> Self;
+
+    /// Normalizes this vector in-place and returns its norm.
+    fn normalize_mut(&mut self) -> S;
+
+    /// Returns a normalized version of this vector unless its norm as smaller or equal to `eps`.
+    fn try_normalize(&self, eps: &S) -> Option<Self>;
+
+    /// Normalizes this vector in-place or does nothing if its norm is smaller or equal to `eps`.
+    ///
+    /// If the normalization succeded, returns the old normal of this vector.
+    fn try_normalize_mut(&mut self, eps: &S) -> Option<S>;
+}
+
+
+/// An approximate vector space aquipped with an inner product.
+///
+/// It must be a normed space as well and the norm must agree with the inner product.
+/// The inner product must be symmetric, linear in its first agurment, and positive definite.
+pub trait InnerSpaceApprox<S: RealApprox>
+    : NormedSpaceApprox<S> {
+    /// Computes the inner product of `self` with `other`.
+    fn inner_product(&self, other: &Self) -> S;
+
+    /// Measures the angle between two vectors.
+    #[inline]
+    fn angle(&self, other: &Self) -> S {
+        let prod = self.inner_product(other);
+        let n1   = self.norm();
+        let n2   = self.norm();
+
+        let zero: S = ident::id(Additive);
+
+        if n1 == Zero::zero() || n2 == Zero::zero() {
+            zero
+        }
+        else {
+            let cang = prod / (n1 * n2);
+
+            cang.acos()
+        }
+    }
+}
+
+/// An approximate finite-dimensional vector space.
+pub trait FiniteDimVectorSpaceApprox<S: FieldApprox>
+    : VectorSpaceApprox<S> {
+
+    /// The vector space dimension.
+    fn dimension() -> usize;
+
+    /// The vector space canonical basis.
+    fn canonical_basis<F: FnOnce(&[Self])>(f: F);
+
+    /// Retrieves the i-th component of `Self` wrt. some basis.
+    ///
+    /// As usual, indexing starts with 0. The actual choice of basis is usually context-dependent
+    /// and is not specified to this method. It is up to the user to assume the provided component
+    /// will by wrt. the suitable basis for his application.
+    fn component(&self, i: usize) -> S;
+
+    /// Same as `.component(i)` but without bound-checking.
+    unsafe fn component_unchecked(&self, i: usize) -> S;
```

Comment:
I wonder about us putting this in here... what about types that have a length that is known statically?